File: backend/app/core/security.py
"""
Security utilities and middleware for LocalGhost API.
"""
import re
import hashlib
import secrets
from typing import Optional, List, Dict, Any
from fastapi import Request, HTTPException, status
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from passlib.context import CryptContext
from app.core.config import settings
import structlog

logger = structlog.get_logger()

# Password hashing context
pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

# Security headers
SECURITY_HEADERS = {
    "X-Content-Type-Options": "nosniff",
    "X-Frame-Options": "DENY",
    "X-XSS-Protection": "1; mode=block",
    "Referrer-Policy": "strict-origin-when-cross-origin",
    "Permissions-Policy": "geolocation=(), microphone=(), camera=()",
    "Content-Security-Policy": "default-src 'self'; script-src 'self' 'unsafe-inline'; style-src 'self' 'unsafe-inline'; img-src 'self' data: https:; font-src 'self' data:; connect-src 'self' https:; frame-ancestors 'none';",
}

class SecurityValidator:
    """Input validation and sanitization utilities."""
    
    # Common attack patterns
    SQL_INJECTION_PATTERNS = [
        r"(\b(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|UNION|SCRIPT)\b)",
        r"(\b(OR|AND)\s+\d+\s*=\s*\d+)",
        r"(\b(OR|AND)\s+'.*'\s*=\s*'.*')",
        r"(--|#|\/\*|\*\/)",
        r"(\b(UNION|UNION ALL)\b)",
    ]
    
    XSS_PATTERNS = [
        r"<script[^>]*>.*?</script>",
        r"javascript:",
        r"on\w+\s*=",
        r"<iframe[^>]*>.*?</iframe>",
        r"<object[^>]*>.*?</object>",
        r"<embed[^>]*>.*?</embed>",
    ]
    
    @classmethod
    def validate_email(cls, email: str) -> bool:
        """Validate email format and security."""
        if not email or len(email) > 254:
            return False
        
        # Basic email regex
        email_pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        if not re.match(email_pattern, email):
            return False
        
        # Check for suspicious patterns
        suspicious_patterns = [
            r'\.{2,}',  # Multiple consecutive dots
            r'@.*@',    # Multiple @ symbols
            r'\.@',     # Dot before @
            r'@\.',     # @ before dot
        ]
        
        for pattern in suspicious_patterns:
            if re.search(pattern, email):
                return False
        
        return True
    
    @classmethod
    def validate_password(cls, password: str) -> Dict[str, Any]:
        """Validate password strength and security."""
        if not password:
            return {"valid": False, "message": "Password is required"}
        
        if len(password) < 8:
            return {"valid": False, "message": "Password must be at least 8 characters long"}
        
        if len(password) > 128:
            return {"valid": False, "message": "Password must be less than 128 characters"}
        
        # Check for common weak patterns
        weak_patterns = [
            r'^[0-9]+$',  # Only numbers
            r'^[a-zA-Z]+$',  # Only letters
            r'^[^a-zA-Z0-9]+$',  # Only special characters
            r'(.)\1{3,}',  # Repeated characters
        ]
        
        for pattern in weak_patterns:
            if re.match(pattern, password):
                return {"valid": False, "message": "Password is too weak"}
        
        # Check for common passwords
        common_passwords = [
            "password", "123456", "123456789", "qwerty", "abc123",
            "password123", "admin", "letmein", "welcome", "monkey"
        ]
        
        if password.lower() in common_passwords:
            return {"valid": False, "message": "Password is too common"}
        
        return {"valid": True, "message": "Password is valid"}
    
    @classmethod
    def sanitize_string(cls, input_string: str, max_length: int = 1000) -> str:
        """Sanitize string input to prevent XSS and injection attacks."""
        if not input_string:
            return ""
        
        # Truncate if too long
        if len(input_string) > max_length:
            input_string = input_string[:max_length]
        
        # Remove null bytes
        input_string = input_string.replace('\x00', '')
        
        # Check for XSS patterns
        for pattern in cls.XSS_PATTERNS:
            if re.search(pattern, input_string, re.IGNORECASE):
                logger.warning("XSS pattern detected in input", pattern=pattern, input=input_string[:100])
                return ""
        
        # Check for SQL injection patterns
        for pattern in cls.SQL_INJECTION_PATTERNS:
            if re.search(pattern, input_string, re.IGNORECASE):
                logger.warning("SQL injection pattern detected in input", pattern=pattern, input=input_string[:100])
                return ""
        
        return input_string.strip()
    
    @classmethod
    def validate_uuid(cls, uuid_string: str) -> bool:
        """Validate UUID format."""
        uuid_pattern = r'^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$'
        return bool(re.match(uuid_pattern, uuid_string, re.IGNORECASE))
    
    @classmethod
    def validate_file_upload(cls, filename: str, content_type: str, file_size: int) -> Dict[str, Any]:
        """Validate file upload security."""
        # Check file size
        if file_size > settings.MAX_FILE_SIZE:
            return {
                "valid": False,
                "message": f"File size exceeds maximum allowed size of {settings.MAX_FILE_SIZE} bytes"
            }
        
        # Check file extension
        allowed_extensions = ['.jpg', '.jpeg', '.png', '.webp', '.gif']
        file_extension = filename.lower().split('.')[-1] if '.' in filename else ''
        
        if f'.{file_extension}' not in allowed_extensions:
            return {
                "valid": False,
                "message": f"File type .{file_extension} is not allowed"
            }
        
        # Check MIME type
        if content_type not in settings.ALLOWED_FILE_TYPES:
            return {
                "valid": False,
                "message": f"Content type {content_type} is not allowed"
            }
        
        # Check for suspicious filename patterns
        suspicious_patterns = [
            r'\.\.',  # Directory traversal
            r'[<>:"|?*]',  # Invalid filename characters
            r'^(CON|PRN|AUX|NUL|COM[1-9]|LPT[1-9])$',  # Windows reserved names
        ]
        
        for pattern in suspicious_patterns:
            if re.search(pattern, filename, re.IGNORECASE):
                return {
                    "valid": False,
                    "message": "Filename contains invalid characters or patterns"
                }
        
        return {"valid": True, "message": "File upload is valid"}

class RateLimiter:
    """Simple in-memory rate limiter."""
    
    def __init__(self):
        self.requests: Dict[str, List[float]] = {}
    
    def is_allowed(self, key: str, limit: int, window: int) -> bool:
        """Check if request is allowed based on rate limit."""
        now = time.time()
        
        # Clean old requests
        if key in self.requests:
            self.requests[key] = [
                req_time for req_time in self.requests[key]
                if now - req_time < window
            ]
        else:
            self.requests[key] = []
        
        # Check if under limit
        if len(self.requests[key]) >= limit:
            return False
        
        # Add current request
        self.requests[key].append(now)
        return True

# Global rate limiter instance
rate_limiter = RateLimiter()

def get_client_ip(request: Request) -> str:
    """Get client IP address from request."""
    # Check for forwarded headers first
    forwarded_for = request.headers.get("X-Forwarded-For")
    if forwarded_for:
        return forwarded_for.split(",")[0].strip()
    
    real_ip = request.headers.get("X-Real-IP")
    if real_ip:
        return real_ip
    
    # Fallback to direct connection
    if request.client:
        return request.client.host
    
    return "unknown"

def hash_password(password: str) -> str:
    """Hash password using bcrypt."""
    return pwd_context.hash(password)

def verify_password(plain_password: str, hashed_password: str) -> bool:
    """Verify password against hash."""
    return pwd_context.verify(plain_password, hashed_password)

def generate_secure_token(length: int = 32) -> str:
    """Generate cryptographically secure random token."""
    return secrets.token_urlsafe(length)

def generate_csrf_token() -> str:
    """Generate CSRF token."""
    return secrets.token_urlsafe(32)

def validate_csrf_token(token: str, session_token: str) -> bool:
    """Validate CSRF token."""
    return secrets.compare_digest(token, session_token)

def sanitize_filename(filename: str) -> str:
    """Sanitize filename for safe storage."""
    # Remove path components
    filename = filename.split('/')[-1].split('\\')[-1]
    
    # Remove or replace dangerous characters
    filename = re.sub(r'[<>:"|?*]', '_', filename)
    
    # Limit length
    if len(filename) > 255:
        name, ext = filename.rsplit('.', 1) if '.' in filename else (filename, '')
        filename = name[:255-len(ext)-1] + ('.' + ext if ext else '')
    
    return filename

def check_rate_limit(request: Request, limit: int = 100, window: int = 60) -> bool:
    """Check if request should be rate limited."""
    client_ip = get_client_ip(request)
    return rate_limiter.is_allowed(client_ip, limit, window)

def validate_request_headers(request: Request) -> bool:
    """Validate request headers for security."""
    # Check for suspicious headers
    suspicious_headers = [
        'X-Forwarded-Host',
        'X-Original-URL',
        'X-Rewrite-URL',
    ]
    
    for header in suspicious_headers:
        if header in request.headers:
            logger.warning("Suspicious header detected", header=header, value=request.headers[header])
            return False
    
    return True

def get_security_headers() -> Dict[str, str]:
    """Get security headers for response."""
    headers = SECURITY_HEADERS.copy()
    
    # Add HSTS in production
    if settings.ENVIRONMENT == "production":
        headers["Strict-Transport-Security"] = "max-age=31536000; includeSubDomains; preload"
    
    return headers

File: frontend/utils/security.ts
/**
 * Frontend security utilities for XSS protection, input validation, and secure data handling.
 */

// XSS Protection
export function sanitizeHtml(input: string): string {
  if (!input) return '';
  
  // Create a temporary div element
  const temp = document.createElement('div');
  temp.textContent = input;
  return temp.innerHTML;
}

export function sanitizeText(input: string): string {
  if (!input) return '';
  
  return input
    .replace(/[<>]/g, '') // Remove < and >
    .replace(/javascript:/gi, '') // Remove javascript: protocol
    .replace(/on\w+\s*=/gi, '') // Remove event handlers
    .trim();
}

export function escapeHtml(input: string): string {
  if (!input) return '';
  
  return input
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#x27;')
    .replace(/\//g, '&#x2F;');
}

// Input Validation
export interface ValidationResult {
  isValid: boolean;
  message?: string;
}

export function validateEmail(email: string): ValidationResult {
  if (!email) {
    return { isValid: false, message: 'Email is required' };
  }
  
  if (email.length > 254) {
    return { isValid: false, message: 'Email is too long' };
  }
  
  const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
  if (!emailRegex.test(email)) {
    return { isValid: false, message: 'Please enter a valid email address' };
  }
  
  // Check for suspicious patterns
  const suspiciousPatterns = [
    /\.{2,}/,  // Multiple consecutive dots
    /@.*@/,    // Multiple @ symbols
    /\.@/,     // Dot before @
    /@\./,     // @ before dot
  ];
  
  for (const pattern of suspiciousPatterns) {
    if (pattern.test(email)) {
      return { isValid: false, message: 'Email contains invalid characters' };
    }
  }
  
  return { isValid: true };
}

export function validatePassword(password: string): ValidationResult {
  if (!password) {
    return { isValid: false, message: 'Password is required' };
  }
  
  if (password.length < 8) {
    return { isValid: false, message: 'Password must be at least 8 characters long' };
  }
  
  if (password.length > 128) {
    return { isValid: false, message: 'Password must be less than 128 characters' };
  }
  
  // Check for common weak patterns
  const weakPatterns = [
    /^[0-9]+$/,  // Only numbers
    /^[a-zA-Z]+$/,  // Only letters
    /^[^a-zA-Z0-9]+$/,  // Only special characters
    /(.)\1{3,}/,  // Repeated characters
  ];
  
  for (const pattern of weakPatterns) {
    if (pattern.test(password)) {
      return { isValid: false, message: 'Password is too weak' };
    }
  }
  
  // Check for common passwords
  const commonPasswords = [
    'password', '123456', '123456789', 'qwerty', 'abc123',
    'password123', 'admin', 'letmein', 'welcome', 'monkey'
  ];
  
  if (commonPasswords.includes(password.toLowerCase())) {
    return { isValid: false, message: 'Password is too common' };
  }
  
  return { isValid: true };
}

export function validateName(name: string): ValidationResult {
  if (!name) {
    return { isValid: false, message: 'Name is required' };
  }
  
  if (name.length < 2) {
    return { isValid: false, message: 'Name must be at least 2 characters long' };
  }
  
  if (name.length > 100) {
    return { isValid: false, message: 'Name must be less than 100 characters' };
  }
  
  // Check for valid characters (letters, spaces, hyphens, apostrophes)
  const nameRegex = /^[a-zA-Z\s\-']+$/;
  if (!nameRegex.test(name)) {
    return { isValid: false, message: 'Name contains invalid characters' };
  }
  
  return { isValid: true };
}

export function validateBio(bio: string): ValidationResult {
  if (!bio) {
    return { isValid: true }; // Bio is optional
  }
  
  if (bio.length > 1000) {
    return { isValid: false, message: 'Bio must be less than 1000 characters' };
  }
  
  // Check for XSS patterns
  const xssPatterns = [
    /<script[^>]*>.*?<\/script>/gi,
    /javascript:/gi,
    /on\w+\s*=/gi,
    /<iframe[^>]*>.*?<\/iframe>/gi,
    /<object[^>]*>.*?<\/object>/gi,
    /<embed[^>]*>.*?<\/embed>/gi,
  ];
  
  for (const pattern of xssPatterns) {
    if (pattern.test(bio)) {
      return { isValid: false, message: 'Bio contains invalid content' };
    }
  }
  
  return { isValid: true };
}

export function validateUrl(url: string): ValidationResult {
  if (!url) {
    return { isValid: true }; // URL is optional
  }
  
  try {
    const urlObj = new URL(url);
    
    // Only allow http and https protocols
    if (!['http:', 'https:'].includes(urlObj.protocol)) {
      return { isValid: false, message: 'Only HTTP and HTTPS URLs are allowed' };
    }
    
    // Check for suspicious patterns
    const suspiciousPatterns = [
      /javascript:/gi,
      /data:/gi,
      /vbscript:/gi,
    ];
    
    for (const pattern of suspiciousPatterns) {
      if (pattern.test(url)) {
        return { isValid: false, message: 'URL contains invalid protocol' };
      }
    }
    
    return { isValid: true };
  } catch {
    return { isValid: false, message: 'Please enter a valid URL' };
  }
}

// File Upload Security
export function validateFileUpload(file: File): ValidationResult {
  if (!file) {
    return { isValid: false, message: 'No file selected' };
  }
  
  // Check file size (10MB limit)
  const maxSize = 10 * 1024 * 1024; // 10MB
  if (file.size > maxSize) {
    return { isValid: false, message: 'File size must be less than 10MB' };
  }
  
  // Check file type
  const allowedTypes = [
    'image/jpeg',
    'image/jpg',
    'image/png',
    'image/webp',
    'image/gif'
  ];
  
  if (!allowedTypes.includes(file.type)) {
    return { isValid: false, message: 'Only image files are allowed' };
  }
  
  // Check file extension
  const allowedExtensions = ['.jpg', '.jpeg', '.png', '.webp', '.gif'];
  const fileExtension = file.name.toLowerCase().split('.').pop();
  
  if (!fileExtension || !allowedExtensions.includes(`.${fileExtension}`)) {
    return { isValid: false, message: 'File type not allowed' };
  }
  
  // Check for suspicious filename patterns
  const suspiciousPatterns = [
    /\.\./,  // Directory traversal
    /[<>:"|?*]/,  // Invalid filename characters
    /^(CON|PRN|AUX|NUL|COM[1-9]|LPT[1-9])$/i,  // Windows reserved names
  ];
  
  for (const pattern of suspiciousPatterns) {
    if (pattern.test(file.name)) {
      return { isValid: false, message: 'Filename contains invalid characters' };
    }
  }
  
  return { isValid: true };
}

// CSRF Protection
export function generateCSRFToken(): string {
  const array = new Uint8Array(32);
  crypto.getRandomValues(array);
  return Array.from(array, byte => byte.toString(16).padStart(2, '0')).join('');
}

export function validateCSRFToken(token: string, sessionToken: string): boolean {
  if (!token || !sessionToken) {
    return false;
  }
  
  // Use timing-safe comparison
  if (token.length !== sessionToken.length) {
    return false;
  }
  
  let result = 0;
  for (let i = 0; i < token.length; i++) {
    result |= token.charCodeAt(i) ^ sessionToken.charCodeAt(i);
  }
  
  return result === 0;
}

// Secure Storage
export function setSecureItem(key: string, value: string): void {
  try {
    // Encrypt sensitive data before storing
    const encrypted = btoa(unescape(encodeURIComponent(value)));
    localStorage.setItem(key, encrypted);
  } catch (error) {
    console.error('Failed to store secure item:', error);
  }
}

export function getSecureItem(key: string): string | null {
  try {
    const encrypted = localStorage.getItem(key);
    if (!encrypted) return null;
    
    // Decrypt data after retrieving
    return decodeURIComponent(escape(atob(encrypted)));
  } catch (error) {
    console.error('Failed to retrieve secure item:', error);
    return null;
  }
}

export function removeSecureItem(key: string): void {
  try {
    localStorage.removeItem(key);
  } catch (error) {
    console.error('Failed to remove secure item:', error);
  }
}

// Content Security Policy
export function createCSPNonce(): string {
  const array = new Uint8Array(16);
  crypto.getRandomValues(array);
  return Array.from(array, byte => byte.toString(16).padStart(2, '0')).join('');
}

// Input Sanitization for Forms
export function sanitizeFormData<T extends Record<string, any>>(data: T): T {
  const sanitized = { ...data };
  
  for (const key in sanitized) {
    if (typeof sanitized[key] === 'string') {
      (sanitized as any)[key] = sanitizeText(sanitized[key] as string);
    }
  }
  
  return sanitized;
}

// URL Validation and Sanitization
export function sanitizeUrl(url: string): string {
  try {
    const urlObj = new URL(url);
    
    // Only allow http and https
    if (!['http:', 'https:'].includes(urlObj.protocol)) {
      return '';
    }
    
    // Remove potentially dangerous parts
    urlObj.search = '';
    urlObj.hash = '';
    
    return urlObj.toString();
  } catch {
    return '';
  }
}

// Phone Number Validation
export function validatePhoneNumber(phone: string): ValidationResult {
  if (!phone) {
    return { isValid: true }; // Phone is optional
  }
  
  // Remove all non-digit characters
  const digits = phone.replace(/\D/g, '');
  
  if (digits.length < 10) {
    return { isValid: false, message: 'Phone number must be at least 10 digits' };
  }
  
  if (digits.length > 15) {
    return { isValid: false, message: 'Phone number must be less than 15 digits' };
  }
  
  return { isValid: true };
}
